<a href="https://colab.research.google.com/github/sohomghosh/FinCausal-2020_2022/blob/master/FinCausal2022_sohom.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
import numpy as np

# Data Preparation

In [ ]:
data_22 = pd.read_csv('task2_2022_addition.csv', sep = ';')
data_22

Index                                               Text  \
0        0  The increase in net interest income was due pr...   
1        1  NPAs increased $703 million year over year, pr...   
2        2  Additional increases in noninterest income wer...   
3        3  During 2020, the Company also issued $6.5 bill...   
4        4  The significant increases in earnings assets a...   
..     ...                                                ...   
533    533  We recorded a gain of $15.7 million for the ye...   
534    534  We recorded a loss of $13.1 million and $3.2 m...   
535    535  We recorded a $4.6 million loss on extinguishm...   
536    536  Accordingly, operating cash flows will be impa...   
537    537  During the year ended December 31, 2020, Wynn ...   

                                                 Cause  \
0    a $152.9 billion increase in average outstandi...   
1    PCI loans that would have been classified as n...   
2    higher insurance income driven by improved pro...   
3    The increase is due to net income in excess of...   
4    the Merger, as well as impacts from the COVID-...   
..                                                 ...   
533  the partial prepayment of the Wynn Macau Term ...   
534  change in the fair value of an interest rate c...   
535  the partial prepayment of the Wynn Macau Term ...   
536  During the year ended December 31, 2020, the d...   
537  the COVID-19 pandemic as noted in Note 1, Orga...   

                                                Effect  
0                  The increase in net interest income  
1           NPAs increased $703 million year over year  
2           Additional increases in noninterest income  
3    During 2020, the Company also issued $6.5 bill...  
4    The significant increases in earnings assets a...  
..                                                 ...  
533  We recorded a $4.6 million loss on extinguishm...  
534  We recorded a loss of $13.1 million and $3.2 m...  
535  We recorded a $4.6 million loss on extinguishm...  
536  Accordingly, operating cash flows will be impa...  
537  During the year ended December 31, 2020, Wynn ...  

[538 rows x 4 columns]

In [ ]:
data_20_task2 = pd.read_csv('fnp2020-fincausal2-task2.csv', sep = ';')
data_20_task2

Index                                               Text  \
0     0001.00005.1   The Sunshine State drew in a net influx of ab...   
1     0001.00005.2   The Sunshine State drew in a net influx of ab...   
2       0001.00007   Florida is unique in that it also draws a lar...   
3       0001.00009   Connecticut, Pennsylvania, New Jersey, Illino...   
4       0001.00011   CLICK HERE TO GET THE FOX BUSINESS APP  Data ...   
...            ...                                                ...   
1104    0585.00024   Outside the benchmark index, Scott Technology...   
1105    0585.00026   QEX Logistics rose 1.5 percent to 70 cents af...   
1106    0588.00003   In other words, if the same prices from last ...   
1107    0588.00007   The producers can lower the price to $80 to r...   
1108    0588.00013   The average ticket price on TickPick, an onli...   

                                                  Cause  \
0      It is consistently one of the most popular de...   
1      It is consistently one of the most popular de...   
2      Florida is unique in that it also draws a lar...   
3              the wealthy were picking up and leaving.   
4      CLICK HERE TO GET THE FOX BUSINESS APP  Data ...   
...                                                 ...   
1104   appointing Pacific Edge executive Kate Rankin...   
1105   telling shareholders that it remains upbeat a...   
1106   In other words, if the same prices from last ...   
1107   The producers can lower the price to $80 to r...   
1108   With multiple theme park attractions, a film ...   

                                                 Effect  Offset_Sentence2  \
0      The Sunshine State drew in a net influx of ab...               159   
1      Florida's $17.7 billion in net AGI dwarves th...               159   
2      more than 85 percent of its net inflow of inc...                     
3      Connecticut, Pennsylvania, New Jersey, Illino...                     
4      New York's outflows to the Sunshine State wer...                     
...                                                 ...               ...   
1104   Outside the benchmark index, Scott Technology...                     
1105         QEX Logistics rose 1.5 percent to 70 cents                     
1106   the $1,095,481 that the play made during the ...                     
1107   As a result of the reductions, the average pr...                91   
1108   The average ticket price on TickPick, an onli...               139   

      Offset_Sentence3   Cause_Start   Cause_End   Effect_Start   Effect_End  \
0                  264           160         264              0          159   
1                  264           160         264            265          421   
2                                  0          90             95          190   
3                                147         187              0          131   
4                                  0         150            152          229   
...                ...           ...         ...            ...          ...   
1104                              78         159              0           71   
1105                              49         179              0           42   
1106                               0          75             82          249   
1107                               0          91             92          248   
1108                             140         308              0          139   

                                               Sentence  
0      <e2>The Sunshine State drew in a net influx o...  
1      The Sunshine State drew in a net influx of ab...  
2      <e1>Florida is unique in that it also draws a...  
3      <e2>Connecticut, Pennsylvania, New Jersey, Il...  
4      <e1>CLICK HERE TO GET THE FOX BUSINESS APP  D...  
...                                                 ...  
1104   <e2>Outside the benchmark index, Scott Techno...  
1105   <e2>QEX Logistics rose 1.5 percent to 70 cent...  
1106   <e1>In other

In [ ]:
data_20_task2_practise = pd.read_csv('practice_fnp2020-fincausal-task2.csv', sep = ';')
data_20_task2_practise

Index                                               Text  \
0     0001.00005.1   The Sunshine State drew in a net influx of ab...   
1     0001.00005.2   The Sunshine State drew in a net influx of ab...   
2       0001.00007   Florida is unique in that it also draws a lar...   
3       0001.00009   Connecticut, Pennsylvania, New Jersey, Illino...   
4       0001.00011   CLICK HERE TO GET THE FOX BUSINESS APP  Data ...   
...            ...                                                ...   
1747    0288.00004  The Liberals also say they would invest $100 m...   
1748    0288.00007  Conservative Leader Andrew Scheer, meanwhile, ...   
1749    0288.00008  The Conservatives say the $900 million measure...   
1750    0288.00012  Work with provinces and territories to complet...   
1751    0289.00006  When an invention or finding is monetised, rev...   

                                                  Cause  \
0      It is consistently one of the most popular de...   
1      It is consistently one of the most popular de...   
2      Florida is unique in that it also draws a lar...   
3              the wealthy were picking up and leaving.   
4      CLICK HERE TO GET THE FOX BUSINESS APP  Data ...   
...                                                 ...   
1747  make sure there are enough qualified workers t...   
1748  anyone who spends between $1,000 and $20,000 o...   
1749  The Conservatives say the $900 million measure...   
1750  Work with provinces and territories to complet...   
1751          When an invention or finding is monetised   

                                                 Effect  Offset_Sentence2  \
0      The Sunshine State drew in a net influx of ab...               159   
1      Florida's $17.7 billion in net AGI dwarves th...               159   
2      more than 85 percent of its net inflow of inc...                     
3      Connecticut, Pennsylvania, New Jersey, Illino...                     
4      New York's outflows to the Sunshine State wer...                     
...                                                 ...               ...   
1747  The Liberals also say they would invest $100 m...               NaN   
1748  Conservative Leader Andrew Scheer, meanwhile, ...               NaN   
1749  Canadians to save  up to $3,800 on renovations...               NaN   
1750  Liberals are offering a $150 million investmen...             123.0   
1751  revenue sharing would be in a 60:40 ratio, wit...               NaN   

      Offset_Sentence3   Cause_Start   Cause_End   Effect_Start   Effect_End  \
0                  264         160.0       264.0            0.0        159.0   
1                  264         160.0       264.0          265.0        421.0   
2                                0.0        90.0           95.0        190.0   
3                              147.0       187.0            0.0        131.0   
4                                0.0       150.0          152.0        229.0   
...                ...           ...         ...            ...          ...   
1747               NaN          75.0       190.0            0.0         71.0   
1748               NaN         188.0       267.0            0.0        162.0   
1749               NaN           0.0        84.0           98.0        156.0   
1750               NaN           0.0       123.0          124.0        218.0   
1751               NaN           0.0        41.0           43.0        211.0   

                                               Sentence  
0      <e2>The Sunshine State drew in a net influx o...  
1      The Sunshine State drew in a net influx of ab...  
2      <e1>Florida is unique in that it also draws a...  
3      <e2>Connecticut, Pennsylvania, New Jersey, Il...  
4      <e1>CLICK HERE TO GET THE FOX BUSINESS APP  D...  
...                                                 ...  
1747  <e2>The Liberals also say they would invest $1...  
1748  <e2>Conservative Leader Andrew Scheer, meanwhi...  
1749  <e1>The Conse

In [ ]:
data_20_task2_trial = pd.read_csv('trial_fnp2020-fincausal-task2.csv', sep = ';')
data_20_task2_trial

Index                                               Text  \
0      0001.00026   It found that total U.S. health care spending...   
1      0003.00001   Transat loss more than doubles as it works to...   
2      0003.00002   MONTREAL  -  Transat AT Inc.'s third-quarter ...   
3      0003.00003   The Montreal-based company lost $11.0 million...   
4      0003.00007   Transatlantic passenger traffic  -  key in th...   
..            ...                                                ...   
636  0363.00001.2   WisdomTree U.S. Quality Dividend Growth Fund ...   
637    0363.00004   NASDAQ:DGRW traded up $0.32 during trading ho...   
638  0364.00001.1   Rural Funds Group (RFF) to Issue Interim Divi...   
639  0364.00001.2   Rural Funds Group (RFF) to Issue Interim Divi...   
640    0364.00003   Shares of Rural Funds Group stock traded down...   

                                                 Cause  \
0     Part of the reason is that Medicare for All w...   
1                 it works to complete Air Canada deal   
2     the tour operator worked to complete its take...   
3     The push to close Transat's sale to Air Canad...   
4     help boost revenue to $698.9 million, up from...   
..                                                 ...   
636   WisdomTree U.S. Quality Dividend Growth Fund ...   
637   NASDAQ:DGRW traded up $0.32 during trading ho...   
638   Rural Funds Group (RFF) to Issue Interim Divi...   
639   Rural Funds Group (RFF) to Issue Interim Divi...   
640   Shares of Rural Funds Group stock traded down...   

                                                Effect  Offset_Sentence2  \
0     It found that total U.S. health care spending...               162   
1                       Transat loss more than doubles                     
2     Transat AT Inc.'s third-quarter net loss more...                     
3     The Montreal-based company lost $11.0 million...               123   
4     Transatlantic passenger traffic  -  key in th...                     
..                                                 ...               ...   
636                   This represents a yield of 2.4%.               139   
637                                    hitting $44.98.                     
638   Stockholders of record on Thursday, October 3...               187   
639         This represents a dividend yield of 1.33%.               187   
640                            hitting A$1.73 ($1.22).                     

     Offset_Sentence3   Cause_Start   Cause_End   Effect_Start   Effect_End  \
0                               163         321              0          162   
1                                34          70              0           30   
2                               104         168             13          100   
3                               124         335              0          123   
4                               177         253              0          173   
..                ...           ...         ...            ...          ...   
636               261             0         139            262          294   
637                               0          61             63           78   
638               307             0         187            188          307   
639               307             0         187            308          350   
640                               0          73             75           98   

                                              Sentence  
0     <e2>It found that total U.S. health care spen...  
1     <e2>Transat loss more than doubles</e2> as <e...  
2     MONTREAL  -  <e2>Transat AT Inc.'s third-quar...  
3     <e2>The Montreal-based company lost $11.0 mil...  
4     <e2>Transatlantic passenger traffic  -  key i...  
..                                                 ...  
636   <e1>WisdomTree U.S. Quality Dividend Growth F...  
637   <e1>NASDAQ:DGRW traded up $0.32 during tradin...  
638   <e1>Rural Funds Group (RFF) to Issue Interim ...  
639   <e1>Rural Fu

In [ ]:
col = [' Text', ' Cause', ' Effect']
data_all = data_22[col].append(data_20_task2[col]).append(data_20_task2_practise[col]).append(data_20_task2_trial[col]).drop_duplicates()
data_all

Text  \
0    The increase in net interest income was due pr...   
1    NPAs increased $703 million year over year, pr...   
2    Additional increases in noninterest income wer...   
3    During 2020, the Company also issued $6.5 bill...   
4    The significant increases in earnings assets a...   
..                                                 ...   
636   WisdomTree U.S. Quality Dividend Growth Fund ...   
637   NASDAQ:DGRW traded up $0.32 during trading ho...   
638   Rural Funds Group (RFF) to Issue Interim Divi...   
639   Rural Funds Group (RFF) to Issue Interim Divi...   
640   Shares of Rural Funds Group stock traded down...   

                                                 Cause  \
0    a $152.9 billion increase in average outstandi...   
1    PCI loans that would have been classified as n...   
2    higher insurance income driven by improved pro...   
3    The increase is due to net income in excess of...   
4    the Merger, as well as impacts from the COVID-...   
..                                                 ...   
636   WisdomTree U.S. Quality Dividend Growth Fund ...   
637   NASDAQ:DGRW traded up $0.32 during trading ho...   
638   Rural Funds Group (RFF) to Issue Interim Divi...   
639   Rural Funds Group (RFF) to Issue Interim Divi...   
640   Shares of Rural Funds Group stock traded down...   

                                                Effect  
0                  The increase in net interest income  
1           NPAs increased $703 million year over year  
2           Additional increases in noninterest income  
3    During 2020, the Company also issued $6.5 bill...  
4    The significant increases in earnings assets a...  
..                                                 ...  
636                   This represents a yield of 2.4%.  
637                                    hitting $44.98.  
638   Stockholders of record on Thursday, October 3...  
639         This represents a dividend yield of 1.33%.  
640                            hitting A$1.73 ($1.22).  

[2813 rows x 3 columns]

In [ ]:
data_all.columns = ['text', 'cause', 'effect']
data_all['index'] = data_all.index

In [ ]:
data_all

In [ ]:
data_all[['index', 'text', 'cause', 'effect']].to_csv('data_all_till2022_dedup.csv', index = False)

# Data Load

In [ ]:
data = pd.read_csv('FINCAUSAL_data_all_till2022_dedup.csv')
data

index                                               text  \
0         0  The increase in net interest income was due pr...   
1         1  NPAs increased $703 million year over year, pr...   
2         2  Additional increases in noninterest income wer...   
3         3  During 2020, the Company also issued $6.5 bill...   
4         4  The significant increases in earnings assets a...   
...     ...                                                ...   
2808    636   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809    637   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810    638   Rural Funds Group (RFF) to Issue Interim Divi...   
2811    639   Rural Funds Group (RFF) to Issue Interim Divi...   
2812    640   Shares of Rural Funds Group stock traded down...   

                                                  cause  \
0     a $152.9 billion increase in average outstandi...   
1     PCI loans that would have been classified as n...   
2     higher insurance income driven by improved pro...   
3     The increase is due to net income in excess of...   
4     the Merger, as well as impacts from the COVID-...   
...                                                 ...   
2808   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810   Rural Funds Group (RFF) to Issue Interim Divi...   
2811   Rural Funds Group (RFF) to Issue Interim Divi...   
2812   Shares of Rural Funds Group stock traded down...   

                                                 effect  
0                   The increase in net interest income  
1            NPAs increased $703 million year over year  
2            Additional increases in noninterest income  
3     During 2020, the Company also issued $6.5 bill...  
4     The significant increases in earnings assets a...  
...                                                 ...  
2808                   This represents a yield of 2.4%.  
2809                                    hitting $44.98.  
2810   Stockholders of record on Thursday, October 3...  
2811         This represents a dividend yield of 1.33%.  
2812                            hitting A$1.73 ($1.22).  

[2813 rows x 4 columns]

In [ ]:
data.sample(60)

index                                               text  \
280     322  As a result of the COVID-19 pandemic and our r...   
1499   1049   The Housing Council, which met on Aug. 21, re...   
282     324  As a result of Grupo Aeroméxico's and LATAM’s ...   
2343    146   It needed a new design that went beyond trans...   
2331    133   Shares of LON:NAIT traded up GBX 3 ($0.04) du...   
1189    735   If rates were negative, the mREIT could be bo...   
1986   1540  - - Sell Ratings - Hold Ratings - Buy Ratings ...   
1903   1457  Vanguard U.S. Quality Factor (VFQY) To Go Ex-D...   
1603   1153  Kolkata: Domestic steel demand and global mark...   
417     490  The remaining increase in nonperforming loans ...   
2373    176   She lowered effective corporate tax further f...   
2153   1707  SPDR Russell 1000 Yield Focus ETF stock traded...   
594     139   Also like 529 savings plans, you'll pay a 10 ...   
1060    605   The firm used $3.5 million in authority funds...   
728     273   Most of the year-to-date revenue shortfall co...   
1815   1366  This should not be going on in America, not fo...   
2118   1672  Materials Select Sector SPDR Fund (NYSEARCA:XL...   
1563   1113  The Brazilian Amazon rainforest is experiencin...   
1771   1321  Guild Investment Management Inc sold 14,510 sh...   
2614    426   Source: InsiderScore.com  First United (FUNC)...   
1857   1411  Vanguard Total International Stock ETF (NASDAQ...   
1286    832   State-run BSNL added an impressive 288,027 mo...   
1522   1072   And finally, OVO's plan to buy SSE's energy b...   
873     418   These companies have not paid taxes worth Rs ...   
2216     18   However, revenue rose 1% to £122.5m. This was...   
2736    557   Economics  The trade war between the United S...   
1783   1333  Lawmakers approved rebate checks earlier this ...   
557     102   In the twelve years of operation there have b...   
2396    199   Again, citing Reuters, Turkey could miss out ...   
2094   1648  Net sales for the quarter ended June 30, 2019,...   
1047    592   Notably, almost all IT firms are below the 25...   
2044   1598  In the short term, the restructuring program i...   
1259    805   Invesco Ltd. boosted its holdings in shares o...   
890     435   After the aborted sale, the French conglomera...   
1698   1248  Dimensional Fund Advisors LP lifted its holdin...   
360     402  The gross profit decline was partially offset ...   
1002    547   Share  September 13, 2019  Rare Daily Staff  ...   
626     171   A breakdown of the hotel's revenues show that...   
134     134  Foreign currency contributed 1.8% to the incre...   
624     169   A breakdown of the hotel's revenues show that...   
1324    870   The limits applicable to individual issues an...   
1782   1332  top story  Letters to the Editor: Give $50 to ...   
889     434   Overstock.com Inc., down 62 cents to $15.57  ...   
1760   1310  Invesco Ltd. raised its stake in shares of Eve...   
1784   1334  We all recognize that our teachers are underpa...   
1078    623   Driven by growth in its Cloud and Surface lap...   
1382    928   In addition, the Asia Pacific region remains ...   
1566   1116  If your property is used for personal and busi...   
2604    416   B.C. strengthened its carbon tax on April 1, ...   
1732   1282  In other First of Long Island news, insider Mi...   
2041   1595  By Current Publishing on Fishers Community  By...   
1008    553   The liquidation of the world's oldest travel ...   
1804   1355  Long-term base fee rates decreased from 18.5 b...   
2176   1730  NYSEARCA DWX traded down $0.14 during trading ...   
2004   1558  The cut in headline rates for existing compani...   
1020    565   Volkswagen has invested USD 100 million in th...   
216     252  In June 2012, the Company completed the offeri...   
968     513   Thomas Cook ceased trading in the early hours...   
294     336  Other charges decreased $34 million, or 15% fr...   
455       0   The Sunshine State drew in a net

# EDA

## Number of words in text, cause, effect

In [ ]:
data['num_words_text'] = data['text'].apply(lambda x : len(str(x).split(' ')))
data['num_words_cause'] = data['cause'].apply(lambda x : len(str(x).split(' ')))
data['num_words_effect'] = data['effect'].apply(lambda x : len(str(x).split(' ')))

In [ ]:
data['num_words_text'].value_counts()

27     71
35     69
30     68
38     67
41     66
       ..
200     1
206     1
158     1
99      1
117     1
Name: num_words_text, Length: 120, dtype: int64

In [ ]:
data['num_words_text'].mean()

43.7045858514042

In [ ]:
data['num_words_cause'].value_counts()

10     127
9      123
11     123
12     117
13     111
      ... 
61       1
74       1
82       1
110      1
62       1
Name: num_words_cause, Length: 84, dtype: int64

In [ ]:
data['num_words_cause'].mean()

19.35407038748667

In [ ]:
data['num_words_effect'].value_counts()

21    129
15    121
6     120
17    119
7     118
     ... 
69      1
94      1
58      1
63      1
64      1
Name: num_words_effect, Length: 67, dtype: int64

In [ ]:
data['num_words_effect'].mean()

17.959829363668682

## How many cases cause longer than effect and vice-versa?

In [ ]:
data[data['num_words_cause']>data['num_words_effect']]

index                                               text  \
0         0  The increase in net interest income was due pr...   
1         1  NPAs increased $703 million year over year, pr...   
2         2  Additional increases in noninterest income wer...   
4         4  The significant increases in earnings assets a...   
6         6  The average rate on long-term debt was 1.75%, ...   
...     ...                                                ...   
2808    636   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809    637   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810    638   Rural Funds Group (RFF) to Issue Interim Divi...   
2811    639   Rural Funds Group (RFF) to Issue Interim Divi...   
2812    640   Shares of Rural Funds Group stock traded down...   

                                                  cause  \
0     a $152.9 billion increase in average outstandi...   
1     PCI loans that would have been classified as n...   
2     higher insurance income driven by improved pro...   
4     the Merger, as well as impacts from the COVID-...   
6     The lower rates on long-term debt also reflect...   
...                                                 ...   
2808   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810   Rural Funds Group (RFF) to Issue Interim Divi...   
2811   Rural Funds Group (RFF) to Issue Interim Divi...   
2812   Shares of Rural Funds Group stock traded down...   

                                                 effect  num_words_text  \
0                   The increase in net interest income              50   
1            NPAs increased $703 million year over year              53   
2            Additional increases in noninterest income              19   
4     The significant increases in earnings assets a...              28   
6     The average rate on long-term debt was 1.75%, ...              84   
...                                                 ...             ...   
2808                   This represents a yield of 2.4%.              48   
2809                                    hitting $44.98.              26   
2810   Stockholders of record on Thursday, October 3...              57   
2811         This represents a dividend yield of 1.33%.              57   
2812                            hitting A$1.73 ($1.22).              24   

      num_words_cause  num_words_effect  
0                  27                 6  
1                  24                 7  
2                  10                 5  
4                  16                 8  
6                  56                17  
...               ...               ...  
2808               22                 7  
2809               10                 3  
2810               30                21  
2811               30                 8  
2812               13                 4  

[1453 rows x 7 columns]

In [ ]:
data[data['num_words_cause']<=data['num_words_effect']].shape[0]

1360

In [ ]:
data = data.dropna()

## Is there overlap between cause & effect? If so, in how many such cases?

In [ ]:
def check_overlap(text, cause, effect):
  text = text.strip()
  cause = cause.strip()
  effect = effect.strip()
  text_without_cause = text.replace(str(cause), '')
  if str(effect) in str(text_without_cause).strip():
    overlap = 0
  else:
    overlap = 1
    print(text, '\n', cause, '\n', text_without_cause, '\n', effect)
  return overlap

In [ ]:
data['overlap'] = data.apply(lambda x : check_overlap(x.text, x.cause, x.effect), axis = 1)

We recorded a gain of $15.7 million for the year ended December 31, 2020 to reflect the fair value of our cost method investment at the date we acquired a controlling interest in BetBull Limited. 
 the partial prepayment of the Wynn Macau Term Loan. 
 We recorded a gain of $15.7 million for the year ended December 31, 2020 to reflect the fair value of our cost method investment at the date we acquired a controlling interest in BetBull Limited. 
 We recorded a $4.6 million loss on extinguishment of debt for the year ended December 31, 2020


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [ ]:
data['overlap'].value_counts()

0    2810
1       1
Name: overlap, dtype: int64

In [ ]:
data[data['overlap']==1]

index                                               text  \
450    533  We recorded a gain of $15.7 million for the ye...   

                                                 cause  \
450  the partial prepayment of the Wynn Macau Term ...   

                                                effect  num_words_text  \
450  We recorded a $4.6 million loss on extinguishm...              35   

     num_words_cause  num_words_effect  overlap  
450                9                17        1

In [ ]:
data.iloc[450,:]

index                                                             535
text                We recorded a $4.6 million loss on extinguishm...
cause               the partial prepayment of the Wynn Macau Term ...
effect              We recorded a $4.6 million loss on extinguishm...
num_words_text                                                     29
num_words_cause                                                     9
num_words_effect                                                   18
overlap                                                             0
Name: 452, dtype: object

In [ ]:
list(data[data['overlap']==1]['text'])

['We recorded a gain of $15.7 million for the year ended December 31, 2020 to reflect the fair value of our cost method investment at the date we acquired a controlling interest in BetBull Limited.']

In [ ]:
list(data[data['overlap']==1]['cause'])

['the partial prepayment of the Wynn Macau Term Loan.']

In [ ]:
list(data[data['overlap']==1]['effect'])

['We recorded a $4.6 million loss on extinguishment of debt for the year ended December 31, 2020']

In [ ]:
'Between the strong and stable dividend yield attracting institutional investors and large and accelerating investment flows into alternative property sectors, such as healthcare and self-storage, analysts are forecasting a doubling of the FTSE EPRA Nareit Developed Europe Index to around EUR 500 bn.'.replace('Between the strong and stable dividend yield attracting institutional investors and large and accelerating investment flows into alternative property sectors, such as healthcare and self-storage', '')#.replace('analysts are forecasting a doubling of the FTSE EPRA Nareit Developed Europe Index to around EUR 500 bn.', '')

', analysts are forecasting a doubling of the FTSE EPRA Nareit Developed Europe Index to around EUR 500 bn.'

**remove and re-create index**

**remove ambiguous records**

In [ ]:
data_use = data[data['overlap']==0].copy()
data_use.shape

(2810, 8)

In [ ]:
data_use['index'] = [i for i in range(2810)]

In [ ]:
data_use[['index', 'text', 'cause', 'effect']]

index                                               text  \
0         0  The increase in net interest income was due pr...   
1         1  NPAs increased $703 million year over year, pr...   
2         2  Additional increases in noninterest income wer...   
3         3  During 2020, the Company also issued $6.5 bill...   
4         4  The significant increases in earnings assets a...   
...     ...                                                ...   
2808   2805   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809   2806   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810   2807   Rural Funds Group (RFF) to Issue Interim Divi...   
2811   2808   Rural Funds Group (RFF) to Issue Interim Divi...   
2812   2809   Shares of Rural Funds Group stock traded down...   

                                                  cause  \
0     a $152.9 billion increase in average outstandi...   
1     PCI loans that would have been classified as n...   
2     higher insurance income driven by improved pro...   
3     The increase is due to net income in excess of...   
4     the Merger, as well as impacts from the COVID-...   
...                                                 ...   
2808   WisdomTree U.S. Quality Dividend Growth Fund ...   
2809   NASDAQ:DGRW traded up $0.32 during trading ho...   
2810   Rural Funds Group (RFF) to Issue Interim Divi...   
2811   Rural Funds Group (RFF) to Issue Interim Divi...   
2812   Shares of Rural Funds Group stock traded down...   

                                                 effect  
0                   The increase in net interest income  
1            NPAs increased $703 million year over year  
2            Additional increases in noninterest income  
3     During 2020, the Company also issued $6.5 bill...  
4     The significant increases in earnings assets a...  
...                                                 ...  
2808                   This represents a yield of 2.4%.  
2809                                    hitting $44.98.  
2810   Stockholders of record on Thursday, October 3...  
2811         This represents a dividend yield of 1.33%.  
2812                            hitting A$1.73 ($1.22).  

[2810 rows x 4 columns]

In [ ]:
data_use[['index', 'text', 'cause', 'effect']].to_csv('FINCAUSAL_data_all_till2022_dedup_clean.csv', index = False)

In [ ]:
def start_posn(text, sub_text):
  text = text.strip()
  sub_text = sub_text.strip()
  try:
    psn = text.index(sub_text)
  except ValueError:
    psn = -1
  return psn

In [ ]:
data_use['start_posn_cause'] = data_use.apply(lambda x : start_posn(x.text, x.cause), axis = 1)

data_use['start_posn_effect'] = data_use.apply(lambda x : start_posn(x.text, x.effect), axis = 1)


In [ ]:
data_use[data_use['start_posn_cause']==-1].shape[0]

0

In [ ]:
data_use[data_use['start_posn_effect']==-1].shape[0]

0

In [ ]:
data_use[data_use['start_posn_cause']==-1]

Empty DataFrame
Columns: [index, text, cause, effect, num_words_text, num_words_cause, num_words_effect, overlap, start_posn_cause, start_posn_effect]
Index: []

In [ ]:
data_use[data_use['start_posn_cause'] < data_use['start_posn_effect']].shape[0]

1716

In [ ]:
data_use[data_use['start_posn_cause'] > data_use['start_posn_effect']].shape[0]

1094

In [ ]:
data_use[data_use['start_posn_cause'] == data_use['start_posn_effect']].shape[0]

0

In [ ]:
data_use['text'].nunique()

2556

In [ ]:
data_use.shape[0]

2810

In [ ]:
2810-2556

254

In [ ]:
df = data_use['text'].value_counts().reset_index()
df.columns = ['text', 'count']
dup_texts = df[df['count']>1]['text'].values
dup_texts

In [ ]:
data_use[data_use['text'].isin(dup_texts)].shape[0]

503

In [ ]:
data_use[data_use['text'].isin(dup_texts)][['index', 'text', 'cause', 'effect']].sort_values('text')

index                                               text  \
513     510   (NASDAQ:JCS) declared a quarterly dividend on...   
514     511   (NASDAQ:JCS) declared a quarterly dividend on...   
548     545   (NASDAQ:JKHY) declared a quarterly dividend o...   
547     544   (NASDAQ:JKHY) declared a quarterly dividend o...   
2389   2386   (NASDAQ:MFNC) announced a 0.12000 dividend on...   
...     ...                                                ...   
1594   1591  Zacks Investment Research cut Kosmos Energy fr...   
1911   1908  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   
1912   1909  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   
1668   1665  iShares Dow Jones US Reg Banks Ind. (BMV:IAT) ...   
1667   1664  iShares Dow Jones US Reg Banks Ind. (BMV:IAT) ...   

                                                  cause  \
513    (NASDAQ:JCS) declared a quarterly dividend on...   
514    (NASDAQ:JCS) declared a quarterly dividend on...   
548    (NASDAQ:JKHY) declared a quarterly dividend o...   
547    (NASDAQ:JKHY) declared a quarterly dividend o...   
2389   (NASDAQ:MFNC) announced a 0.12000 dividend on...   
...                                                 ...   
1594  Zacks Investment Research cut Kosmos Energy fr...   
1911  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   
1912  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   
1668  iShares Dow Jones US Reg Banks Ind. (BMV:IAT) ...   
1667  iShares Dow Jones US Reg Banks Ind. (BMV:IAT) ...   

                                                 effect  
513    Investors of record on Monday, September 16th...  
514    This represents a $0.08 dividend on an annual...  
548    This represents a $1.60 annualized dividend a...  
547    Investors of record on Monday, September 9th ...  
2389   Investors of record on Friday, September 27th...  
...                                                 ...  
1594  Kosmos Energy has an average rating of Buy and...  
1911  Investors of record on Wednesday, September 25...  
1912                   This represents a yield of 1.5%.  
1668                   This represents a yield of 2.4%.  
1667  Shareholders of record on Wednesday, September...  

[503 rows x 4 columns]

In [ ]:
data_use[data_use['text'].isin(dup_texts)]['text'].nunique()

249

In [ ]:
data_use[data_use['text'].isin(dup_texts)][['text', 'cause']].drop_duplicates().shape[0]

352

In [ ]:
data_use[data_use['text'].isin(dup_texts)][['text', 'cause', 'effect']].drop_duplicates().shape[0]

503

In [ ]:
data_use[data_use['text'].isin(dup_texts)][['text', 'effect']].drop_duplicates().shape[0]

446

In [ ]:
data_use[data_use['text'].isin(dup_texts)][['text', 'cause']].value_counts().reset_index()

text  \
0     Photo: Creativa - Fotolia  Health insurers mu...   
1     (NASDAQ:JCS) declared a quarterly dividend on...   
2     Sells 2,011 Shares of BlackRock Enhanced Glob...   
3     Since the 1990s, China has made numerous effo...   
4     So you discover that both micro, micro plus a...   
..                                                 ...   
347   They issued an equal weight rating and a $74....   
348   They issued an equal weight rating and a $74....   
349   They set a neutral rating and a $147.00 targe...   
350   They set a neutral rating and a $147.00 targe...   
351   Shadow business secretary Rebecca Long Bailey...   

                                                 cause  0  
0     The mandate to issue refunds comes under a pr...  3  
1     (NASDAQ:JCS) declared a quarterly dividend on...  2  
2     Sells 2,011 Shares of BlackRock Enhanced Glob...  2  
3     Since the 1990s, China has made numerous effo...  2  
4     So you discover that both micro, micro plus a...  2  
..                                                 ... ..  
347   They issued an equal weight rating and a $74....  1  
348   Two research analysts have rated the stock wi...  1  
349   Four research analysts have rated the stock w...  1  
350   They set a neutral rating and a $147.00 targe...  1  
351                               save the company now  1  

[352 rows x 3 columns]

In [ ]:
dfc = data_use[data_use['text'].isin(dup_texts)][['text', 'cause']].value_counts().reset_index()
dfc.columns = ['text', 'cause', 'count']
dfc[dfc['count']>1]

text  \
0     Photo: Creativa - Fotolia  Health insurers mu...   
1     (NASDAQ:JCS) declared a quarterly dividend on...   
2     Sells 2,011 Shares of BlackRock Enhanced Glob...   
3     Since the 1990s, China has made numerous effo...   
4     So you discover that both micro, micro plus a...   
..                                                 ...   
145   But the QBI deduction changed that thinking. ...   
146   IQ 50 Percent Hedged FTSE Europe ETF (NYSEARC...   
147   But all in all, Orbán's program, universally ...   
148   (NASDAQ:MFNC) announced a 0.12000 dividend on...   
149   Announces Quarterly Dividend of $0.06 (NASDAQ...   

                                                 cause  count  
0     The mandate to issue refunds comes under a pr...      3  
1     (NASDAQ:JCS) declared a quarterly dividend on...      2  
2     Sells 2,011 Shares of BlackRock Enhanced Glob...      2  
3     Since the 1990s, China has made numerous effo...      2  
4     So you discover that both micro, micro plus a...      2  
..                                                 ...    ...  
145       But the QBI deduction changed that thinking.      2  
146   IQ 50 Percent Hedged FTSE Europe ETF (NYSEARC...      2  
147   But all in all, Orbán's program, universally ...      2  
148   (NASDAQ:MFNC) announced a 0.12000 dividend on...      2  
149   Announces Quarterly Dividend of $0.06 (NASDAQ...      2  

[150 rows x 3 columns]

In [ ]:
dfe = data_use[data_use['text'].isin(dup_texts)][['text', 'effect']].value_counts().reset_index()
dfe.columns = ['text', 'effect', 'count']
dfe[dfe['count']>1]

text  \
0    Central banks, led by those in China and Russ...   
1   The $57.2 million year-over-year increase in o...   
2    Finally, Morgan Stanley set a $21.00 price ta...   
3    Finally, Loop Capital upgraded Cinemark from ...   
4    Finally, KeyCorp increased their target price...   
5    Finally, Deutsche Bank set a $33.00 price obj...   
6    Finally, Deutsche Bank reaffirmed a sell rati...   
7    Finally, Buckingham Research cut their price ...   
8    Finally, Bank of America reduced their price ...   
9    Finally, BMO Capital Markets set a $28.00 pri...   
10  As sentiment soured, partly owing to WeWork's ...   
11   Earlier this year it signed up for a $6 billi...   
12  It recorded a 4.8 per cent slide in people's e...   
13  LHFS decreased $2.3 billion during 2020 primar...   
14  Macroeconomic conditions, including Brexit wor...   
15  Finally, CIBC lowered Mercer International fro...   
16   But after 12 months of welcoming drivers, fig...   
17   You get a bonus of 2% of your account balance...   
18   Based on geography, China and Japan countries...   
19   Finally, Needham & Company LLC reissued a buy...   
20   Finally, TD Securities boosted their price ta...   
21   Finally, ValuEngine downgraded Hersha Hospita...   
22  Finally, BidaskClub lowered shares of News fro...   
23   India is a crucial and an integral part of ou...   
24  Finally, ValuEngine cut shares of Gladstone Co...   
25  Finally, ValuEngine downgraded shares of Globa...   
26  Finally, ValuEngine upgraded Ovid Therapeutics...   
27  Finally, Zacks Investment Research raised Houl...   
28  Finally, Zacks Investment Research upgraded sh...   
29  Finally, CIBC boosted their price objective on...   
30  Global Economic Developments Output growth acr...   
31   Finally, ValuEngine lowered shares of Travele...   
32  Finally, BidaskClub cut Sunrun from a buy rati...   
33   It will be held to re-engage with the market ...   
34   Let's begin with a story: At the end of 2017,...   
35  Berenberg Bank raised Smith & Nephew from a ho...   
36   Finally, Zacks Investment Research upgraded E...   
37   Finally, Zacks Investment Research raised sha...   
38   Finally, Zacks Investment Research raised Con...   
39   As for their separate arguments, Fraser-Jenki...   
40  Finally, Credit Suisse Group upgraded shares o...   
41  The increase in net interest income was due pr...   
42   We think that, right now, private money is be...   
43  They set an outperform rating and a $97.00 pri...   
44   Shares have risen 44 percent over the past ye...   
45   Two grandparents, Joanie and Bill Miller from...   
46   ASX-listed Sims Metal has warned escalating t...   
47   They set a neutral rating and a $147.00 targe...   
48   They issued an equal weight rating and a $74....   
49   An expert committee that submitted a report o...   
50   Analysts expect Lockheed Martin to report FY ...   
51   Sandler O'Neill lowered shares of Customers B...   
52  They set an overweight rating and a $122.00 pr...   
53  Zacks Investment Research cut Kosmos Energy fr...   
54   The index showed that Singaporeans earning mi...   
55   The government will provide a 25% bonus on th...   
56   The final result for the 2018/19 financial ye...   

                                               effect  count  
0    Russia has quadrupled its gold reserves in th...      2  
1   The $57.2 million year-over-year increase in o...      2  
2    The company currently has a consensus rating ...      2  
3    The stock presently has a consensus rating of...      2  
4    The stock currently has a consensus rating of...      2  
5    The company presently has an average rating o...      2  
6    The company currently has a consensus rating ...      2  
7    The stock currently has a consensus rating of...      2  
8    The company presently has a consensus rating ...      2  
9    The company presently has a consensus rating ...      2  
10  reports surfaced of Mr Neumann contemplating 

In [ ]:
set(dfc[dfc['count']>1]['text'].values).intersection(set(dfe[dfe['count']>1]['text'].values))

{" Two grandparents, Joanie and Bill Miller from Buffalo, New York, decided to make Apple their Single-Stock Retirement Plan back in 1997. They put $16,000 into Apple stock. And since then, they've made $2 million."}

In [ ]:
len(set(dfc[dfc['count']>1]['text'].values).intersection(set(dfe[dfe['count']>1]['text'].values)))

1

In [ ]:
sn = " Two grandparents, Joanie and Bill Miller from Buffalo, New York, decided to make Apple their Single-Stock Retirement Plan back in 1997. They put $16,000 into Apple stock. And since then, they've made $2 million."
data_use[data_use['text']==sn]

index                                               text  \
2259   2256   Two grandparents, Joanie and Bill Miller from...   
2260   2257   Two grandparents, Joanie and Bill Miller from...   
2261   2258   Two grandparents, Joanie and Bill Miller from...   

                                                  cause  \
2259   Two grandparents, Joanie and Bill Miller from...   
2260   Two grandparents, Joanie and Bill Miller from...   
2261                 They put $16,000 into Apple stock.   

                                         effect  num_words_text  \
2259   And since then, they've made $2 million.              35   
2260         They put $16,000 into Apple stock.              35   
2261   And since then, they've made $2 million.              35   

      num_words_cause  num_words_effect  overlap  
2259               22                 8        0  
2260               22                 7        0  
2261                7                 8        0

In [ ]:
data_use[data_use['text'].isin(dup_texts) & data_use['cause'].isin(dup_causes) & data_use['effect'].isin(dup_effects)].shape[0]

26

In [ ]:
data_use[data_use['text'].isin(dup_texts) & data_use['cause'].isin(dup_causes) & data_use['effect'].isin(dup_effects)][['text', 'cause', 'effect']].sort_values(['text', 'cause', 'effect'])

text  \
2390   (NASDAQ:MFNC) announced a 0.12000 dividend on...   
2346   Games Workshop Group PLC (GAW) to Issue Divid...   
2347   Games Workshop Group PLC (GAW) to Issue Divid...   
2623   Games Workshop Group PLC (LON:GAW) Raises Div...   
2624   Games Workshop Group PLC (LON:GAW) Raises Div...   
2715   IQ 50 Percent Hedged FTSE Europe ETF (NYSEARC...   
2580   Invesco S&P 500 ex-Rate Sensitive Low Volatil...   
2724   Materials Select Sector SPDR Fund (NYSEARCA:X...   
2733   Photo: Creativa - Fotolia  Health insurers mu...   
2734   Photo: Creativa - Fotolia  Health insurers mu...   
1472   Refusing to do so is ideological shortsighted...   
2751   SPDR S&P Homebuilders ETF (NYSEARCA:XHB) decl...   
1052   Shadow business secretary Rebecca Long Bailey...   
2713   The business also recently declared a monthly...   
1152   The company also recently declared a monthly ...   
903    The mandate to issue refunds comes under a pr...   
904    The mandate to issue refunds comes under a pr...   
2259   Two grandparents, Joanie and Bill Miller from...   
1474   WisdomTree U.S. Total Market Fund (NYSEARCA:E...   
850    iShares MSCI USA Equal Weighted ETF (NYSEARCA...   
480    iShares S&P Mid-Cap 400 Value ETF (NYSEARCA:I...   
1230   iShares U.S. Aerospace & Defense ETF (BATS:IT...   
1684  Invesco PureBeta MSCI USA ETF (BATS:PBUS) anno...   
1629  Invesco Russell 1000 Low Beta Equal Weight ETF...   
2098  Smith & Nephew plc (SNN) to Issue  -  Dividend...   
1912  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   

                                                  cause  \
2390   (NASDAQ:MFNC) announced a 0.12000 dividend on...   
2346   Games Workshop Group PLC (GAW) to Issue Divid...   
2347   Games Workshop Group PLC (GAW) to Issue Divid...   
2623   Games Workshop Group PLC (LON:GAW) Raises Div...   
2624   Games Workshop Group PLC (LON:GAW) Raises Div...   
2715   IQ 50 Percent Hedged FTSE Europe ETF (NYSEARC...   
2580   Invesco S&P 500 ex-Rate Sensitive Low Volatil...   
2724   Materials Select Sector SPDR Fund (NYSEARCA:X...   
2733   The mandate to issue refunds comes under a pr...   
2734   The mandate to issue refunds comes under a pr...   
1472                       repatriate UK holidaymakers.   
2751   SPDR S&P Homebuilders ETF (NYSEARCA:XHB) decl...   
1052                       repatriate UK holidaymakers.   
2713   The business also recently declared a monthly...   
1152   The company also recently declared a monthly ...   
903    The mandate to issue refunds comes under a pr...   
904    The mandate to issue refunds comes under a pr...   
2259   Two grandparents, Joanie and Bill Miller from...   
1474   WisdomTree U.S. Total Market Fund (NYSEARCA:E...   
850    iShares MSCI USA Equal Weighted ETF (NYSEARCA...   
480    iShares S&P Mid-Cap 400 Value ETF (NYSEARCA:I...   
1230   iShares U.S. Aerospace & Defense ETF (BATS:IT...   
1684  Invesco PureBeta MSCI USA ETF (BATS:PBUS) anno...   
1629  Invesco Russell 1000 Low Beta Equal Weight ETF...   
2098  Smith & Nephew plc (SNN) to Issue  -  Dividend...   
1912  iShares Core S&P Mid-Cap ETF (NYSEARCA:IJH) to...   

                                                 effect  
2390                   This represents a yield of 3.1%.  
2346   Stockholders of record on Thursday, September...  
2347         This represents a dividend yield of 0.73%.  
2623   Stockholders of record on Thursday, September...  
2624         This represents a dividend yield of 0.73%.  
2715                   This represents a yield of 3.1%.  
2580                   This represents a yield of 1.7%.  
2724                     This represents a yield of 2%.  
2733   The bulk of the rebates  -  a total of $80.4 ...  
2734   The rest, roughly $11.5 million, will be refu...  
1472                                a £600 million bill  
2751                     This represents a yield of 1%.  
1052                                a £600 million bill  
2713   This represents a $0.76 dividend on an annua

# Observations
* In a text 'effect' may occur before 'cause'. For 1716 cases 'cause' is occurring before 'effect'. For 1094 cases 'effect' is occuring before 'cause'.
* No cases with overlaping 'effect' and 'cause' with each other
* 503 instances have 'text' which are repeating
* These 503 cases have 249 unique 'text', 352 unique 'cause' and 446 unique 'effect'. This indicates the many to many relationships between 'cause' and 'effect'. However, these many to many relationships exists mostly in different instances / rows / 'texts'. Only one case is there where the many-to-many relationship exists in a single 'text'.

# Modeling

In [ ]:
final_data = pd.read_csv('FINCAUSAL_data_all_till2022_dedup_clean.csv')
final_data

index                                               text  \
0         0  The increase in net interest income was due pr...   
1         1  NPAs increased $703 million year over year, pr...   
2         2  Additional increases in noninterest income wer...   
3         3  During 2020, the Company also issued $6.5 bill...   
4         4  The significant increases in earnings assets a...   
...     ...                                                ...   
2805   2805   WisdomTree U.S. Quality Dividend Growth Fund ...   
2806   2806   NASDAQ:DGRW traded up $0.32 during trading ho...   
2807   2807   Rural Funds Group (RFF) to Issue Interim Divi...   
2808   2808   Rural Funds Group (RFF) to Issue Interim Divi...   
2809   2809   Shares of Rural Funds Group stock traded down...   

                                                  cause  \
0     a $152.9 billion increase in average outstandi...   
1     PCI loans that would have been classified as n...   
2     higher insurance income driven by improved pro...   
3     The increase is due to net income in excess of...   
4     the Merger, as well as impacts from the COVID-...   
...                                                 ...   
2805   WisdomTree U.S. Quality Dividend Growth Fund ...   
2806   NASDAQ:DGRW traded up $0.32 during trading ho...   
2807   Rural Funds Group (RFF) to Issue Interim Divi...   
2808   Rural Funds Group (RFF) to Issue Interim Divi...   
2809   Shares of Rural Funds Group stock traded down...   

                                                 effect  
0                   The increase in net interest income  
1            NPAs increased $703 million year over year  
2            Additional increases in noninterest income  
3     During 2020, the Company also issued $6.5 bill...  
4     The significant increases in earnings assets a...  
...                                                 ...  
2805                   This represents a yield of 2.4%.  
2806                                    hitting $44.98.  
2807   Stockholders of record on Thursday, October 3...  
2808         This represents a dividend yield of 1.33%.  
2809                            hitting A$1.73 ($1.22).  

[2810 rows x 4 columns]

In [ ]:
np.random.seed(0)
msk = np.random.rand(final_data.shape[0]) < 0.8

In [ ]:
# pd.DataFrame({'index':final_data.index,'is_train':msk}).to_csv('train_valid_split.csv', index = False)

In [ ]:
train = final_data[msk]
valid = final_data[~msk]

## Exploring BERT for Token classification

https://huggingface.co/docs/transformers/model_doc/bert

In [ ]:
!pip install transformers

In [ ]:
from transformers import BertTokenizer, BertForTokenClassification
import torch

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertForTokenClassification.from_pretrained("bert-base-uncased")

inputs = tokenizer("Hello, my dog is cute", return_tensors="pt")
labels = torch.tensor([1] * inputs["input_ids"].size(1)).unsqueeze(0)  # Batch size 1

outputs = model(**inputs, labels=labels)
loss = outputs.loss
logits = outputs.logits

Downloading:   0%|          | 0.00/226k [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/28.0 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/570 [00:00<?, ?B/s]

Downloading:   0%|          | 0.00/420M [00:00<?, ?B/s]

Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForTokenClassification: ['cls.predictions.transform.LayerNorm.weight', 'cls.seq_relationship.bias', 'cls.predictions.transform.LayerNorm.bias', 'cls.predictions.transform.dense.weight', 'cls.predictions.bias', 'cls.predictions.transform.dense.bias', 'cls.predictions.decoder.weight', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForTokenClassification from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForTokenClassification from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Some weights of BertForTokenClassification were not initialized from the model checkpoint at bert-base-u

In [ ]:
inputs

{'input_ids': tensor([[  101,  7592,  1010,  2026,  3899,  2003, 10140,   102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 1]])}

In [ ]:
labels

tensor([[1, 1, 1, 1, 1, 1, 1, 1]])

In [ ]:
outputs

TokenClassifierOutput([('loss', tensor(0.6832, grad_fn=<NllLossBackward0>)),
                       ('logits', tensor([[[-0.0217, -0.5220],
                                 [ 0.0435, -0.2303],
                                 [-0.1990, -0.1151],
                                 [-0.1429,  0.1384],
                                 [-0.4002,  0.1498],
                                 [ 0.0934,  0.1949],
                                 [ 0.0754,  0.0787],
                                 [ 0.1496,  0.2436]]], grad_fn=<AddBackward0>))])

In [ ]:
loss

tensor(0.6832, grad_fn=<NllLossBackward0>)

In [ ]:
logits

tensor([[[-0.0217, -0.5220],
         [ 0.0435, -0.2303],
         [-0.1990, -0.1151],
         [-0.1429,  0.1384],
         [-0.4002,  0.1498],
         [ 0.0934,  0.1949],
         [ 0.0754,  0.0787],
         [ 0.1496,  0.2436]]], grad_fn=<AddBackward0>)

In [ ]:
# https://colab.research.google.com/github/NielsRogge/Transformers-Tutorials/blob/master/BERT/Custom_Named_Entity_Recognition_with_BERT_only_first_wordpiece.ipynb

In [ ]:
!pip install seqeval[gpu]

     |████████████████████████████████| 43 kB 1.1 MB/s 
  Created wheel for seqeval: filename=seqeval-1.2.2-py3-none-any.whl size=16180 sha256=17d6fcca42fa01b73db44a352fe2f2c7f48f21fffa9ecc8260da643b8353e44a
  Stored in directory: /root/.cache/pip/wheels/05/96/ee/7cac4e74f3b19e3158dce26a20a1c86b3533c43ec72a549fd7
Successfully built seqeval


In [ ]:
!pip install transformers

     |████████████████████████████████| 3.8 MB 5.0 MB/s 
     |████████████████████████████████| 895 kB 60.7 MB/s 
     |████████████████████████████████| 596 kB 43.9 MB/s 
     |████████████████████████████████| 6.5 MB 40.4 MB/s 
     |████████████████████████████████| 67 kB 4.9 MB/s 
  Attempting uninstall: pyyaml
    Found existing installation: PyYAML 3.13
    Uninstalling PyYAML-3.13:
      Successfully uninstalled PyYAML-3.13


In [ ]:
import pandas as pd
import numpy as np
from sklearn.metrics import accuracy_score
import torch
from torch.utils.data import Dataset, DataLoader
from transformers import BertTokenizerFast, BertConfig, BertForTokenClassification

In [ ]:
from torch import cuda
device = 'cuda' if cuda.is_available() else 'cpu'
print(device)

cpu


In [ ]:
data # should have two columns 
#sentence;	word_labels

#Thousands of demonstrators have marched throug...;	O,O,O,O,O,O,B-geo,O,O,O,O,O,B-geo,O,O,O,O,O,B-...
# B-C = Begin Cause
# I-C = Inside Cause
# B-E = Begin Effect
# I-E = Inside Effect
# O = Outside


In [ ]:
final_data

index                                               text  \
0         0  The increase in net interest income was due pr...   
1         1  NPAs increased $703 million year over year, pr...   
2         2  Additional increases in noninterest income wer...   
3         3  During 2020, the Company also issued $6.5 bill...   
4         4  The significant increases in earnings assets a...   
...     ...                                                ...   
2805   2805   WisdomTree U.S. Quality Dividend Growth Fund ...   
2806   2806   NASDAQ:DGRW traded up $0.32 during trading ho...   
2807   2807   Rural Funds Group (RFF) to Issue Interim Divi...   
2808   2808   Rural Funds Group (RFF) to Issue Interim Divi...   
2809   2809   Shares of Rural Funds Group stock traded down...   

                                                  cause  \
0     a $152.9 billion increase in average outstandi...   
1     PCI loans that would have been classified as n...   
2     higher insurance income driven by improved pro...   
3     The increase is due to net income in excess of...   
4     the Merger, as well as impacts from the COVID-...   
...                                                 ...   
2805   WisdomTree U.S. Quality Dividend Growth Fund ...   
2806   NASDAQ:DGRW traded up $0.32 during trading ho...   
2807   Rural Funds Group (RFF) to Issue Interim Divi...   
2808   Rural Funds Group (RFF) to Issue Interim Divi...   
2809   Shares of Rural Funds Group stock traded down...   

                                                 effect  
0                   The increase in net interest income  
1            NPAs increased $703 million year over year  
2            Additional increases in noninterest income  
3     During 2020, the Company also issued $6.5 bill...  
4     The significant increases in earnings assets a...  
...                                                 ...  
2805                   This represents a yield of 2.4%.  
2806                                    hitting $44.98.  
2807   Stockholders of record on Thursday, October 3...  
2808         This represents a dividend yield of 1.33%.  
2809                            hitting A$1.73 ($1.22).  

[2810 rows x 4 columns]

In [ ]:
# target
#sentence	word_labels
#Thousands of demonstrators have marched throug...	O,O,O,O,O,O,B-geo,O,O,O,O,O,B-geo,O,O,O,O,O,B-...

In [ ]:
def token_tag_assign(text, cause, effect):
  #only_text = text.replace(cause, ' ').replace(effect, ' ')
  cause_index = text.index(cause)
  effect_index = text.index(effect)
  wds = []
  tks = []
  if cause_index<=effect_index:
    text_before_cause = text[0:cause_index]
    #cause
    text_bw_cause_effect= text[cause_index:effect_index]
    #effect
    text_after_effect= text[effect_index:]

    for wd in text_before_cause.split(" "):
      wds.append(wd)
      tks.append("O")

    wds.append(cause.split(" ")[0])
    tks.append("B-C")
    for wd in cause.split(" ")[1:]:
      wds.append(wd)
      tks.append("I-C")

    for wd in text_bw_cause_effect.split(" "):
      wds.append(wd)
      tks.append("O")

    wds.append(effect.split(" ")[0])
    tks.append("B-E")
    for wd in effect.split(" ")[1:]:
      wds.append(wd)
      tks.append("I-E")
    
    for wd in text_after_effect.split(" "):
      wds.append(wd)
      tks.append("O")

  else:
    text_before_effect = text[0:effect_index]
    #effect
    text_bw_effect_cause= text[effect_index:cause_index]
    #cause
    text_after_cause= text[cause_index:]

    


In [ ]:
MAX_LEN = 128
TRAIN_BATCH_SIZE = 4
VALID_BATCH_SIZE = 2
EPOCHS = 1
LEARNING_RATE = 1e-05
MAX_GRAD_NORM = 10
tokenizer = BertTokenizerFast.from_pretrained('bert-base-uncased')

In [ ]:
class dataset(Dataset):
  def __init__(self, dataframe, tokenizer, max_len):
        self.len = len(dataframe)
        self.data = dataframe
        self.tokenizer = tokenizer
        self.max_len = max_len

  def __getitem__(self, index):
        # step 1: get the sentence and word labels 
        sentence = self.data.sentence[index].strip().split()  
        word_labels = self.data.word_labels[index].split(",") 

        # step 2: use tokenizer to encode sentence (includes padding/truncation up to max length)
        # BertTokenizerFast provides a handy "return_offsets_mapping" functionality for individual tokens
        encoding = self.tokenizer(sentence,
                             is_pretokenized=True, 
                             return_offsets_mapping=True, 
                             padding='max_length', 
                             truncation=True, 
                             max_length=self.max_len)
        
        # step 3: create token labels only for first word pieces of each tokenized word
        labels = [labels_to_ids[label] for label in word_labels] 
        # code based on https://huggingface.co/transformers/custom_datasets.html#tok-ner
        # create an empty array of -100 of length max_length
        encoded_labels = np.ones(len(encoding["offset_mapping"]), dtype=int) * -100
        
        # set only labels whose first offset position is 0 and the second is not 0
        i = 0
        for idx, mapping in enumerate(encoding["offset_mapping"]):
          if mapping[0] == 0 and mapping[1] != 0:
            # overwrite label
            encoded_labels[idx] = labels[i]
            i += 1

        # step 4: turn everything into PyTorch tensors
        item = {key: torch.as_tensor(val) for key, val in encoding.items()}
        item['labels'] = torch.as_tensor(encoded_labels)
        
        return item

  def __len__(self):
        return self.len

In [ ]:
train_size = 0.8
train_dataset = data.sample(frac=train_size,random_state=200)
test_dataset = data.drop(train_dataset.index).reset_index(drop=True)
train_dataset = train_dataset.reset_index(drop=True)

print("FULL Dataset: {}".format(data.shape))
print("TRAIN Dataset: {}".format(train_dataset.shape))
print("TEST Dataset: {}".format(test_dataset.shape))

training_set = dataset(train_dataset, tokenizer, MAX_LEN)
testing_set = dataset(test_dataset, tokenizer, MAX_LEN)

In [ ]:
model = BertForTokenClassification.from_pretrained('bert-base-uncased', num_labels=len(labels_to_ids))
model.to(device)

In [ ]:
inputs = training_set[2]
input_ids = inputs["input_ids"].unsqueeze(0)
attention_mask = inputs["attention_mask"].unsqueeze(0)
labels = inputs["labels"].unsqueeze(0)

input_ids = input_ids.to(device)
attention_mask = attention_mask.to(device)
labels = labels.to(device)

outputs = model(input_ids, attention_mask=attention_mask, labels=labels)
initial_loss = outputs[0]

In [ ]:
tr_logits = outputs[1]

In [ ]:
optimizer = torch.optim.Adam(params=model.parameters(), lr=LEARNING_RATE)

In [ ]:
# Defining the training function on the 80% of the dataset for tuning the bert model
def train(epoch):
    tr_loss, tr_accuracy = 0, 0
    nb_tr_examples, nb_tr_steps = 0, 0
    tr_preds, tr_labels = [], []
    # put model in training mode
    model.train()
    
    for idx, batch in enumerate(training_loader):
        
        ids = batch['input_ids'].to(device, dtype = torch.long)
        mask = batch['attention_mask'].to(device, dtype = torch.long)
        labels = batch['labels'].to(device, dtype = torch.long)

        loss, tr_logits = model(input_ids=ids, attention_mask=mask, labels=labels)
        tr_loss += loss.item()

        nb_tr_steps += 1
        nb_tr_examples += labels.size(0)
        
        if idx % 100==0:
            loss_step = tr_loss/nb_tr_steps
            print(f"Training loss per 100 training steps: {loss_step}")
           
        # compute training accuracy
        flattened_targets = labels.view(-1) # shape (batch_size * seq_len,)
        active_logits = tr_logits.view(-1, model.num_labels) # shape (batch_size * seq_len, num_labels)
        flattened_predictions = torch.argmax(active_logits, axis=1) # shape (batch_size * seq_len,)
        
        # only compute accuracy at active labels
        active_accuracy = labels.view(-1) != -100 # shape (batch_size, seq_len)
        #active_labels = torch.where(active_accuracy, labels.view(-1), torch.tensor(-100).type_as(labels))
        
        labels = torch.masked_select(flattened_targets, active_accuracy)
        predictions = torch.masked_select(flattened_predictions, active_accuracy)
        
        tr_labels.extend(labels)
        tr_preds.extend(predictions)

        tmp_tr_accuracy = accuracy_score(labels.cpu().numpy(), predictions.cpu().numpy())
        tr_accuracy += tmp_tr_accuracy
    
        # gradient clipping
        torch.nn.utils.clip_grad_norm_(
            parameters=model.parameters(), max_norm=MAX_GRAD_NORM
        )
        
        # backward pass
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

    epoch_loss = tr_loss / nb_tr_steps
    tr_accuracy = tr_accuracy / nb_tr_steps
    print(f"Training loss epoch: {epoch_loss}")
    print(f"Training accuracy epoch: {tr_accuracy}")

In [ ]:
for epoch in range(EPOCHS):
    print(f"Training epoch: {epoch + 1}")
    train(epoch)

In [ ]:
def valid(model, testing_loader):
    # put model in evaluation mode
    model.eval()
    
    eval_loss, eval_accuracy = 0, 0
    nb_eval_examples, nb_eval_steps = 0, 0
    eval_preds, eval_labels = [], []
    
    with torch.no_grad():
        for idx, batch in enumerate(testing_loader):
            
            ids = batch['input_ids'].to(device, dtype = torch.long)
            mask = batch['attention_mask'].to(device, dtype = torch.long)
            labels = batch['labels'].to(device, dtype = torch.long)
            
            loss, eval_logits = model(input_ids=ids, attention_mask=mask, labels=labels)
            
            eval_loss += loss.item()

            nb_eval_steps += 1
            nb_eval_examples += labels.size(0)
        
            if idx % 100==0:
                loss_step = eval_loss/nb_eval_steps
                print(f"Validation loss per 100 evaluation steps: {loss_step}")
              
            # compute evaluation accuracy
            flattened_targets = labels.view(-1) # shape (batch_size * seq_len,)
            active_logits = eval_logits.view(-1, model.num_labels) # shape (batch_size * seq_len, num_labels)
            flattened_predictions = torch.argmax(active_logits, axis=1) # shape (batch_size * seq_len,)
            
            # only compute accuracy at active labels
            active_accuracy = labels.view(-1) != -100 # shape (batch_size, seq_len)
        
            labels = torch.masked_select(flattened_targets, active_accuracy)
            predictions = torch.masked_select(flattened_predictions, active_accuracy)
            
            eval_labels.extend(labels)
            eval_preds.extend(predictions)
            
            tmp_eval_accuracy = accuracy_score(labels.cpu().numpy(), predictions.cpu().numpy())
            eval_accuracy += tmp_eval_accuracy

    labels = [ids_to_labels[id.item()] for id in eval_labels]
    predictions = [ids_to_labels[id.item()] for id in eval_preds]
    
    eval_loss = eval_loss / nb_eval_steps
    eval_accuracy = eval_accuracy / nb_eval_steps
    print(f"Validation Loss: {eval_loss}")
    print(f"Validation Accuracy: {eval_accuracy}")

    return labels, predictions

In [ ]:
labels, predictions = valid(model, testing_loader)

In [ ]:
from seqeval.metrics import classification_report

print(classification_report(labels, predictions))

In [ ]:
sentence = "@HuggingFace is a company based in New York, but is also has employees working in Paris"

inputs = tokenizer(sentence.split(),
                    is_pretokenized=True, 
                    return_offsets_mapping=True, 
                    padding='max_length', 
                    truncation=True, 
                    max_length=MAX_LEN,
                    return_tensors="pt")

# move to gpu
ids = inputs["input_ids"].to(device)
mask = inputs["attention_mask"].to(device)
# forward pass
outputs = model(ids, attention_mask=mask)
logits = outputs[0]

active_logits = logits.view(-1, model.num_labels) # shape (batch_size * seq_len, num_labels)
flattened_predictions = torch.argmax(active_logits, axis=1) # shape (batch_size*seq_len,) - predictions at the token level

tokens = tokenizer.convert_ids_to_tokens(ids.squeeze().tolist())
token_predictions = [ids_to_labels[i] for i in flattened_predictions.cpu().numpy()]
wp_preds = list(zip(tokens, token_predictions)) # list of tuples. Each tuple = (wordpiece, prediction)

prediction = []
for token_pred, mapping in zip(wp_preds, inputs["offset_mapping"].squeeze().tolist()):
  #only predictions on first word pieces are important
  if mapping[0] == 0 and mapping[1] != 0:
    prediction.append(token_pred[1])
  else:
    continue

print(sentence.split())
print(prediction)

In [ ]:
import os

directory = "./model"

if not os.path.exists(directory):
    os.makedirs(directory)

# save vocabulary of the tokenizer
tokenizer.save_vocabulary(directory)
# save the model weights and its configuration file
model.save_pretrained(directory)
print('All files saved')
print('This tutorial is completed')

In [ ]:
def prepare_sentence(sentence, tokenizer, maxlen):    
      # step 1: tokenize the sentence
      tokenized_sentence = tokenizer.tokenize(sentence)
      
      # step 2: add special tokens 
      tokenized_sentence = ["[CLS]"] + tokenized_sentence + ["[SEP]"] 

      # step 3: truncating/padding
      if (len(tokenized_sentence) > maxlen):
        # truncate
        tokenized_sentence = tokenized_sentence[:maxlen]
      else:
        # pad
        tokenized_sentence = tokenized_sentence + ['[PAD]'for _ in range(maxlen - len(tokenized_sentence))]

      # step 4: obtain the attention mask
      attn_mask = [1 if tok != '[PAD]' else 0 for tok in tokenized_sentence]
      
      # step 5: convert tokens to input ids
      ids = tokenizer.convert_tokens_to_ids(tokenized_sentence)
      
      return {
            'ids': torch.tensor(ids, dtype=torch.long),
            'mask': torch.tensor(attn_mask, dtype=torch.long),
            #'token_type_ids': torch.tensor(token_ids, dtype=torch.long),
      }

In [ ]:
def tokenize_and_preserve_labels(sentence, text_labels, tokenizer):
    """
    Word piece tokenization makes it difficult to match word labels
    back up with individual word pieces. This function tokenizes each
    word one at a time so that it is easier to preserve the correct
    label for each subword. It is, of course, a bit slower in processing
    time, but it will help our model achieve higher accuracy.
    """

    tokenized_sentence = []
    labels = []

    sentence = sentence.strip()

    for word, label in zip(sentence.split(), text_labels.split(",")):

        # Tokenize the word and count # of subwords the word is broken into
        tokenized_word = tokenizer.tokenize(word)
        n_subwords = len(tokenized_word)

        # Add the tokenized word to the final tokenized word list
        tokenized_sentence.extend(tokenized_word)

        # Add the same label to the new list of labels `n_subwords` times
        labels.extend([label] * n_subwords)

    return tokenized_sentence, labels

In [ ]:
class dataset(Dataset):
    def __init__(self, dataframe, tokenizer, max_len):
        self.len = len(dataframe)
        self.data = dataframe
        self.tokenizer = tokenizer
        self.max_len = max_len
        
    def __getitem__(self, index):
        # step 1: tokenize (and adapt corresponding labels)
        sentence = self.data.sentence[index]  
        word_labels = self.data.word_labels[index]  
        tokenized_sentence, labels = tokenize_and_preserve_labels(sentence, word_labels, self.tokenizer)
        
        # step 2: add special tokens (and corresponding labels)
        tokenized_sentence = ["[CLS]"] + tokenized_sentence + ["[SEP]"] # add special tokens
        labels.insert(0, "O") # add outside label for [CLS] token
        labels.insert(-1, "O") # add outside label for [SEP] token

        # step 3: truncating/padding
        maxlen = self.max_len

        if (len(tokenized_sentence) > maxlen):
          # truncate
          tokenized_sentence = tokenized_sentence[:maxlen]
          labels = labels[:maxlen]
        else:
          # pad
          tokenized_sentence = tokenized_sentence + ['[PAD]'for _ in range(maxlen - len(tokenized_sentence))]
          labels = labels + ["O" for _ in range(maxlen - len(labels))]

        # step 4: obtain the attention mask
        attn_mask = [1 if tok != '[PAD]' else 0 for tok in tokenized_sentence]
        
        # step 5: convert tokens to input ids
        ids = self.tokenizer.convert_tokens_to_ids(tokenized_sentence)

        label_ids = [labels_to_ids[label] for label in labels]
        # the following line is deprecated
        #label_ids = [label if label != 0 else -100 for label in label_ids]
        
        return {
              'ids': torch.tensor(ids, dtype=torch.long),
              'mask': torch.tensor(attn_mask, dtype=torch.long),
              #'token_type_ids': torch.tensor(token_ids, dtype=torch.long),
              'targets': torch.tensor(label_ids, dtype=torch.long)
        } 
    
    def __len__(self):
        return self.len

In [ ]:
sentence = "this is a test @huggingface".strip().split()

inputs = tokenizer(sentence, is_pretokenized=True, return_offsets_mapping=True, padding='max_length', truncation=True)
tokens = tokenizer.convert_ids_to_tokens(inputs["input_ids"])
token_offsets = inputs["offset_mapping"]
print(tokens)
print(token_offsets)

In [ ]:
word = "@huggingface"

inputs = tokenizer(word, return_offsets_mapping=True, padding='max_length', truncation=True)
tokens = tokenizer.convert_ids_to_tokens(inputs["input_ids"])
token_offsets = inputs["offset_mapping"]
print(tokens)
print(token_offsets)

In [ ]:
# now, use mask to determine where we should compare predictions with targets (includes [CLS] and [SEP] token predictions)
        active_accuracy = mask.view(-1) == 1 # active accuracy is also of shape (batch_size * seq_len,)
        targets = torch.masked_select(flattened_targets, active_accuracy)
        predictions = torch.masked_select(flattened_predictions, active_accuracy)